src/System.Text.JsonLab/System/Text/Json/Utf8JsonReader_MultiSegment.cs

```diff
@@ -0,0 +1,438 @@
+            ValueType = JsonValueType.Number;
+            GetNumber(Value);
+        }
+
+        private void ConsumeNull(ref BufferReader<byte> reader)
+        {
+            Value = JsonConstants.NullValue;
+            ValueType = JsonValueType.Null;
+
+            if (!reader.IsNext(JsonConstants.NullValue, advancePast: true))
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+            CurrentIndex += 4;
+        }
+
+        private void ConsumeFalse(ref BufferReader<byte> reader)
+        {
+            Value = JsonConstants.FalseValue;
+            ValueType = JsonValueType.False;
+
+            if (!reader.IsNext(JsonConstants.FalseValue, advancePast: true))
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+            CurrentIndex += 5;
+        }
+
+        private void ConsumeTrue(ref BufferReader<byte> reader)
+        {
+            Value = JsonConstants.TrueValue;
+            ValueType = JsonValueType.True;
+
+            if (!reader.IsNext(JsonConstants.TrueValue, advancePast: true))
+            {
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+            CurrentIndex += 4;
+        }
+
+        private void ConsumePropertyName(ref BufferReader<byte> reader)
+        {
+            ConsumeString(ref reader);
+
+            if (!reader.TryPeek(out byte first))
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+
+            if (first <= JsonConstants.Space)
+            {
+                SkipWhiteSpace(ref reader);
+                if (!reader.TryPeek(out first))
+                    JsonThrowHelper.ThrowJsonReaderException(ref this);
+            }
+
+            // The next character must be a key / value seperator. Validate and skip.
+            if (first != JsonConstants.KeyValueSeperator)
+                JsonThrowHelper.ThrowJsonReaderException(ref this);
+
+            TokenType = JsonTokenType.PropertyName;
+            reader.Advance(1);
+            CurrentIndex++;
+        }
+
+        private void ConsumeString(ref BufferReader<byte> reader)
+        {
+            ReadOnlySpan<byte> buffer = reader.UnreadSpan;
+            int idx = buffer.IndexOf(JsonConstants.Quote);
+            if (idx < 0)
+            {
+                ConsumeStringUtf8MultiSegmentSlow(ref reader);
+                return;
+            }
+
+            if (idx == 0 || buffer[idx - 1] != JsonConstants.ReverseSolidus)
+            {
+                Value = buffer.Slice(0, idx);
+                ValueType = JsonValueType.String;
+
+                idx++;
+                reader.Advance(idx);
+                CurrentIndex += idx;
+                return;
+            }
+            ConsumeStringWithNestedQuotes(ref reader);
+        }
+
+        private void ConsumeStringWithNestedQuotes(ref BufferReader<byte> reader)
+        {
+            //TODO: Optimize looking for nested quotes
```

Comment:
Do we have an API on BufferReader that can read to end of a string? If not, should we add it? If yes, does it support escape sequences?

Reply:
We don't. `TryReadTo` can do scenarios that are in the first 128 chars (ASCII), which would be the most common need I'd think. You could compose off of `TryReadToAny` for escaping or we could provide a method for ASCII escape chars? Something like:

``` C#
/// <param name="escape">When encountered, skip the following {T}.</param>
public bool TryReadTo(out ReadOnlySpan<T> span, T delimiter, T escape, bool advancePastDelimiter = true)
```

Reply:
Other escape chars are really rare so perhaps we just support backslash.


``` C#
public bool TryReadToSkipEscaped(out ReadOnlySpan<T> span, T delimiter, bool advancePastDelimiter = true)
```

Reply:
Such an API would definitely be useful for JSON reading.